File: src/interaction/SquareWell.cpp
/*
  Copyright (C) 2012,2013
  Max Planck Institute for Polymer Research
  Copyright (C) 2008,2009,2010,2011
  Max-Planck-Institute for Polymer Research & Fraunhofer SCAI

  This file is part of ESPResSo++.

  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "python.hpp"
#include "SquareWell.hpp"
#include "VerletListInteractionTemplate.hpp"
#include "FixedPairListInteractionTemplate.hpp"

namespace espressopp {
  namespace interaction {

    typedef class VerletListInteractionTemplate < SquareWell > VerletListSquareWell;
    typedef class FixedPairListInteractionTemplate< SquareWell > FixedPairListSquareWell;
    typedef class FixedPairListTypesInteractionTemplate< SquareWell > FixedPairTypesSquareWell;

    //////////////////////////////////////////////////
    // REGISTRATION WITH PYTHON
    //////////////////////////////////////////////////
    void SquareWell::registerPython() {
      using namespace espressopp::python;

      class_ <SquareWell, bases <Potential> >
    	("interaction_SquareWell", init< real, real, real >())
        .def(init< real, real, real, real >())
        .add_property("epsilon", &SquareWell::getEpsilon, &SquareWell::setEpsilon)
        .add_property("sigma", &SquareWell::getSigma, &SquareWell::setSigma)
        .add_property("width", &SquareWell::getLambda, &SquareWell::setLambda)
        .add_property("a", &SquareWell::getA, &SquareWell::setA)
        .def_pickle(SquareWell_pickle())
        ;

      class_ <VerletListSquareWell, bases <Interaction> >
        ("interaction_VerletListSquareWell", init< shared_ptr<VerletList> >())
        .def("getVerletList", &VerletListSquareWell::getVerletList)
        .def("setPotential", &VerletListSquareWell::setPotential, return_value_policy< reference_existing_object >())
        .def("getPotential", &VerletListSquareWell::getPotential, return_value_policy< reference_existing_object >())
        ;

      class_ <FixedPairListSquareWell, bases <Interaction> >
        ("interaction_FixedPairListSquareWell",
         init< shared_ptr<System>, shared_ptr<FixedPairList>,  shared_ptr<SquareWell> >())
        .def(init< shared_ptr<System>, shared_ptr<FixedPairListAdress>, shared_ptr<SquareWell> >())
        .def("setPotential", &FixedPairListSquareWell::setPotential)
        .def("setFixedPairList", &FixedPairListSquareWell::setFixedPairList)
        .def("getFixedPairList", &FixedPairListSquareWell::getFixedPairList)
        ;
    }
  }
}

File: src/interaction/SquareWell.py
#  Copyright (C) 2012,2013,2018
#      Max Planck Institute for Polymer Research
#  Copyright (C) 2008,2009,2010,2011
#      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
#
#  This file is part of ESPResSo++.
#
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

from espressopp import pmi, infinity
from espressopp.esutil import *

from espressopp.interaction.Potential import *
from espressopp.interaction.Interaction import *
from _espressopp import interaction_SquareWell, interaction_VerletListSquareWell, \
    interaction_FixedPairListSquareWell

class SquareWellLocal(PotentialLocal, interaction_SquareWell):

    def __init__(self, epsilon=1.0, sigma=0.0,
                 cutoff=infinity, shift=0.0):
        """Initialize the local Harmonic object."""
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            if shift == "auto":
                cxxinit(self, interaction_SquareWell, epsilon, sigma, cutoff)
            else:
                cxxinit(self, interaction_SquareWell, epsilon, sigma, cutoff, shift)

class VerletListSquareWellLocal(InteractionLocal, interaction_VerletListSquareWell):

    def __init__(self, vl):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, interaction_VerletListSquareWell, vl)

    def setPotential(self, type1, type2, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setPotential(self, type1, type2, potential)

    def getPotential(self, type1, type2):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            return self.cxxclass.getPotential(self, type1, type2)

    def getVerletListLocal(self):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            return self.cxxclass.getVerletList(self)

class FixedPairListSquareWellLocal(InteractionLocal, interaction_FixedPairListSquareWell):

    def __init__(self, system, vl, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, interaction_FixedPairListSquareWell, system, vl, potential)

    def setPotential(self, potential):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setPotential(self, potential)

    def setFixedPairList(self, fixedpairlist):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            self.cxxclass.setFixedPairList(self, fixedpairlist)

    def getFixedPairList(self):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            return self.cxxclass.getFixedPairList(self)

if pmi.isController:
    class SquareWell(Potential):
        'The SquareWell potential.'
        pmiproxydefs = dict(
            cls = 'espressopp.interaction.SquareWellLocal',
            pmiproperty = ['epsilon', 'sigma', 'width', 'a']
        )

    class VerletListSquareWell(Interaction):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls = 'espressopp.interaction.VerletListSquareWellLocal',
            pmicall = ['setPotential', 'getPotential', 'getVerletList']
        )

    class FixedPairListSquareWell(Interaction):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls =  'espressopp.interaction.FixedPairListSquareWellLocal',
            pmicall = ['setPotential','setFixedPairList','getFixedPairList']
            )
